File: Go/quick_sort/quick_sort.go
package main

import "fmt"

func main() {
	a := []int{9, 12, 3, 6, 4, 0, 3, 22, 7}

	fmt.Println("Unordered: ", a)
	quickSort(&a)
	fmt.Println("Ordered: ", a)
}

func quickSort(a *[]int) {
	quick_sort(a, 0, len(*a))
}

func quick_sort(a *[]int, start int, end int) {
	for end-start > 1 {
		m := partition(a, start, end)
		if m-start < end-m {
			quick_sort(a, start, m)
			start = m + 1
		} else {
			quick_sort(a, m+1, end)
			end = m
		}
	}
}

func partition(a *[]int, start int, end int) int {
	pivot := (*a)[start]
	sx := start
	dx := end

	for sx < dx {

		for ok := true; ok; ok = (*a)[dx] > pivot {
			dx -= 1
		}

		for ok := true; ok; ok = sx < dx && (*a)[sx] <= pivot {
			sx += 1
		}

		if sx < dx {
			swap(a, sx, dx)
		}
	}

	swap(a, start, dx)
	return dx
}

func swap(a *[]int, from int, to int) {
	tmp := (*a)[from]
	(*a)[from] = (*a)[to]
	(*a)[to] = tmp
}

File: C/quick_sort/quick_sort.c
#include <stdio.h>

#define SWAP(a, from, to)                                                      \
  int tmp = a[from];                                                           \
  a[from] = a[to];                                                             \
  a[to] = tmp

#define PRINT_A(a, size_a)                                                     \
  printf("[ ");                                                                \
  for (int i = 0; i < size_a; i++) {                                           \
    printf("%d ", a[i]);                                                       \
  }                                                                            \
  printf("]\n");

int partition(int *a, int start, int end);
void m_quick_sort(int *a, int start, int end);
void quick_sort(int *a, int size_a);

int main() {
  int a[] = {8, 20, 12, 9, 11, 42, 51, 3, 2, 35};
  int size_a = 10;
  
  printf("Unordered: ");
  PRINT_A(a, size_a);
  quick_sort(a, size_a);
  printf("Ordered: ");
  PRINT_A(a, size_a);
    
  return 1;
}

void quick_sort(int *a, int size_a) { m_quick_sort(a, 0, size_a); }

int partition(int *a, int start, int end) {
  int pivot = a[start];
  int sx = start;
  int dx = end;
  while (sx < dx) {
    do {
      dx -= 1;
    } while (a[dx] > pivot);

    do {
      sx += 1;
    } while (sx < dx && a[sx] <= pivot);

    if (sx < dx) {
      SWAP(a, sx, dx);
    }
  }
  SWAP(a, start, dx);
  return dx;
}

void m_quick_sort(int *a, int start, int end) {
  while (end - start > 1) {
    int m = partition(a, start, end);
    if (m - start < end - m) {
      m_quick_sort(a, start, m);
      start = m + 1;
    } else {
      m_quick_sort(a, m + 1, end);
      end = m;
    }
  }
}
